### PR TITLE
feat(chat): Implement ability to use custom renderers for the attachments(Issue #1270)

### PR DESCRIPTION
**Pull request by denys-kolomiitsev**
**Description:**

Visualizer connector lib to connect custom visualizer for the custom content type

Issues:

- Issue #1270 

**Checklist:**

- [x] the pull request name complies with [Conventional Commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] the pull request name starts with `fix(<scope>):`, `feat(<scope>):`, `feature(<scope>):`, `chore(<scope>):`, `hotfix(<scope>):` or `e2e(<scope>):`. If contains breaking changes then the pull request name must start with `fix(<scope>)!:`, `feat(<scope>)!:`, `feature(<scope>)!:`, `chore(<scope>)!:`, `hotfix(<scope>)!:` or `e2e(<scope>)!:` where `<scope>` is name of affected project: `chat`, `chat-e2e`, `overlay`, `shared`, `sandbox-overlay`, etc.
- [x] the pull request name ends with `(Issue #<TICKET_ID>)` (comma-separated list of issues)
- [x] I confirm that do not share any confidential information like API keys or any other secrets and private URLs

### CODE REVIEW

File: apps/chat/src/components/Chat/MessageAttachment.tsx [hunk 1]
```diff
@@ -183,6 +186,48 @@ interface Props {
...
+  const customRenderers = useAppSelector(
+    SettingsSelectors.selectCustomRenderers,
+  );
+  const customAttachmentsTypes = useAppSelector(
+    SettingsSelectors.selectCustomAttachmentsTypes,
+  );
+
+  const mappedRenderers = useMemo(() => {
+    const renderers: Record<string, CustomRenderer[]> = {};
+    customRenderers?.forEach((rendererConfig) => {
```
Comment (mikitabut) [inline apps/chat/src/components/Chat/MessageAttachment.tsx]: nit: I think it's better to use reduce here

File: apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+  }, [Title, rendererUrl]);
+
+  useEffect(() => {
+    if (ready && visualizer.current) {
+      sendMessage(visualizer.current);
+    }
+  }, [ready, attachmentUrl, mimeType, sendMessage]);
+
+  useEffect(() => {
+    const postMessageListener = (event: MessageEvent<any>) => {
```
Comment (mikitabut) [inline apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx]: fix type here

File: apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    };
+
+    window.addEventListener('message', postMessageListener, false);
+
+    return () => window.removeEventListener('message', postMessageListener);
+  }, [Title, rendererUrl]);
+
+  return (
+    <div ref={iframeContainerRef}>
+      <h1>{Title}</h1>
```
Comment (mikitabut) [inline apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx]: It should not be h1 - you will break SEO for dial chat in such case

File: apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+        visualizerData,
+      });
+    },
+    [mimeType],
+  );
+
+  useEffect(() => {
+    if (!visualizer.current) {
+      visualizer.current = new VisualizerConnector(
+        iframeContainerRef.current!,
```
Comment (mikitabut) [inline apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx]: It's better to check for iframeContainerRef.current in top if

File: apps/chat/environment.d.ts [hunk 5]
```diff
@@ -11,6 +11,9 @@ declare global {
       APP_BASE_ORIGIN?: string;
       ALLOWED_IFRAME_ORIGINS?: string;
       IS_IFRAME?: string;
+      ALLOWED_VISUALIZERS_IFRAME_ORIGINS?: string;
+      CUSTOM_VISUALIZERS?: string;
+      CUSTOM_CONTENT_TYPES?: string;
```
Comment (mikitabut) [inline apps/chat/environment.d.ts]: Maybe it's better to have name like CUSTOM_VISUALIZERS_CONTENT_TYPES

File: libs/visualizer-connector/package.json [hunk 6]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "@epam/ai-dial-visualizer-connector",
+  "description": "Package connect custom visualizer",
```
Comment (mikitabut) [inline libs/visualizer-connector/package.json]: it's better to add some words about dial like "Package for connecting custom visualizer into Dial Chat"

File: libs/visualizer-connector/src/index.ts [hunk 7]
```diff
@@ -0,0 +1,3 @@
+export * from './lib/VisualizerConnector';
+
+export { Feature, type VisualizerConnectorOptions } from '@epam/ai-dial-shared';
```
Comment (mikitabut) [inline libs/visualizer-connector/src/index.ts]: Why do you sharing Feature if it's not needed when using this new lib?

File: libs/visualizer-connector/src/lib/VisualizerConnector.ts [hunk 8]
```diff
@@ -0,0 +1,329 @@
...
+<path opacity="0.6" d="M24 1.5V9" stroke="#7F8792" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/>
+<path opacity="0.7" d="M39.9121 8.08594L37.2607 10.7373L34.6094 13.3887" stroke="#7F8792" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/>
+<path opacity="0.8" d="M46.5 24H39" stroke="#7F8792" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/>
+<path opacity="0.9" d="M34.6055 34.6055L39.9082 39.9082" stroke="#7F8792" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/>
+<path opacity="0.3" d="M13.3936 34.6055L8.08594 39.9131" stroke="#7F8792" stroke-width="3" stroke-linecap="round" stroke-linejoin="round"/>
+<animateTransform attributeName="transform" attributeType="XML" type="rotate" from="0 0 0" to="360 0 0" dur="1s" repeatCount="indefinite" />
+</svg>
+`;
+
+export interface VisualizerOptions {
```
Comment (mikitabut) [inline libs/visualizer-connector/src/lib/VisualizerConnector.ts]: Do you need this interface here? Below I can see that you are using VisualizerConnectorOptions

File: libs/visualizer-connector/src/lib/utils/DeferredRequest.ts
Comment (mikitabut) [inline libs/visualizer-connector/src/lib/utils/DeferredRequest.ts:1]: If this is the same Deffered request as in Overlay lib move it to shared lib and reuse

File: libs/visualizer-connector/src/lib/utils/Task.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
+export class Task {
```
Comment (mikitabut) [inline libs/visualizer-connector/src/lib/utils/Task.ts]: Same about Task

File: libs/visualizer-connector/src/lib/utils/styleUtils.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
+import { Styles } from '@epam/ai-dial-shared';
```
Comment (mikitabut) [inline libs/visualizer-connector/src/lib/utils/styleUtils.ts]: same about this

File: libs/visualizer-connector/src/lib/VisualizerConnector.ts [hunk 11]
```diff
@@ -0,0 +1,329 @@
...
+  protected requests: DeferredRequest[];
+
+  protected options: VisualizerConnectorOptions;
+
+  /**
+   * Creates a VisualizerConnector
+   * @param root {HTMLElement | string} reference or selector to parent container where the iframe should be placed
+   * @param options {VisualizerConnectorOptions} visualizer options ( hostDomain, etc.)
+   */
+  constructor(root: HTMLElement | string, options: VisualizerConnectorOptions) {
```
Comment (mikitabut) [inline libs/visualizer-connector/src/lib/VisualizerConnector.ts:50]: It's not clear why are you having 2 set of options and when I should one or another
Looks like some mistake

File: apps/chat/src/store/settings/settings.reducers.ts [hunk 12]
```diff
@@ -37,6 +40,8 @@ const initialState: SettingsState = {
   defaultRecentAddonsIds: [],
   storageType: StorageType.BrowserStorage,
   themesHostDefined: false,
+  customRenderers: [],
+  customAttachmentsTypes: ['image/png'],
```
Comment (mikitabut) [inline apps/chat/src/store/settings/settings.reducers.ts]: I think this is mistake? This is should not be custom attachment type

File: apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+        {
+          domain: rendererUrl,
+          hostDomain: window.location.origin,
+          visualizerName: Title,
+        },
+      );
+    }
+  }, [Title, rendererUrl]);
```
Comment (mikitabut) [inline apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx]: Looks like you forgot to destroy  the connector

File: libs/visualizer-connector/src/lib/VisualizerConnector.ts [hunk 14]
```diff
@@ -0,0 +1,332 @@
...
+
+  /**
+   * Creates iframe add set initial options to it
+   * @returns {HTMLIFrameElement} reference to iframe element
+   */
+  protected initIframe(): HTMLIFrameElement {
+    const iframe = document.createElement('iframe');
+
+    iframe.src = this.options.domain;
+    iframe.allow = 'clipboard-write';
```
Comment (akryvasheyeu) [inline libs/visualizer-connector/src/lib/VisualizerConnector.ts]: I think we grant a lot by default

File: tsconfig.json [hunk 15]
```diff
@@ -17,6 +17,7 @@
     "paths": {
       "@epam/ai-dial-overlay": ["libs/overlay/src/index.ts"],
       "@epam/ai-dial-shared": ["libs/shared/src/index.ts"],
+      "@epam/visualizer-connector": ["libs/visualizer-connector/src/index.ts"],
```
Comment (akryvasheyeu) [inline tsconfig.json]: please use the same unique name `@epam/ai-dial-visualizer-connector` everywhere

File: apps/chat/environment.d.ts [hunk 16]
```diff
@@ -11,6 +11,9 @@ declare global {
       APP_BASE_ORIGIN?: string;
       ALLOWED_IFRAME_ORIGINS?: string;
       IS_IFRAME?: string;
+      ALLOWED_VISUALIZERS_IFRAME_ORIGINS?: string;
```
Comment (akryvasheyeu) [inline apps/chat/environment.d.ts]: I think we should choose a more generic name

File: apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx [hunk 17]
```diff
@@ -0,0 +1,98 @@
...
+      visualizer.send(VisualizerConnectorRequests.sendVisualizeData, {
+        mimeType,
+        visualizerData,
+      });
+    },
+    [mimeType],
+  );
+
+  useEffect(() => {
+    if (iframeContainerRef.current && !visualizer.current) {
```
Comment (akryvasheyeu) [inline apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx:51]: very good place for bugs in future

File: apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx [hunk 18]
```diff
@@ -0,0 +1,98 @@
...
+
+  const [ready, setReady] = useState<boolean>();
+  const { Url: rendererUrl, Title } = renderer;
+
+  const sendMessage = useCallback(
+    async (visualizer: VisualizerConnector) => {
+      await visualizer.ready();
+
+      const visualizerData = {
+        message: 'Hello, I am a message from the chat!',
```
Comment (akryvasheyeu) [inline apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx]: why do we use this?

File: apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx [hunk 19]
```diff
@@ -0,0 +1,98 @@
...
+
+  useEffect(() => {
+    const postMessageListener = (
+      event: MessageEvent<VisualizerConnectorRequest>,
+    ) => {
+      if (event.origin !== rendererUrl) return;
+
+      if (
+        event.data.type ===
+        `${Title}/${VisualizerConnectorEvents.readyToInteract}`
```
Comment (akryvasheyeu) [inline apps/chat/src/components/VisualalizerRenderer/VisualizerRenderer.tsx]: can we reuse one renderer multiple times or should each one have a unique name in our configuration?

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1297.nightly-test.deltixhub.io
>E2E tests status: failed

File: libs/visualizer-connector/README.md [hunk 20]
```diff
@@ -0,0 +1,193 @@
...
+  type: VisualizerConnectorRequests;
+  requestId: string;
+  payload: AttachmentData | VisualizerOptions;
+}
+
+interface PostMessageRequestParams {
+  requestId: string;
+  dialHost: string;
+  payload?: unknown;
+}
```
Comment (IlyaBondar) [inline libs/visualizer-connector/README.md]: All of this typing should be available from our library, shouldn't it?

File: apps/chat/environment.d.ts [hunk 21]
```diff
@@ -11,6 +11,9 @@ declare global {
       APP_BASE_ORIGIN?: string;
       ALLOWED_IFRAME_ORIGINS?: string;
       IS_IFRAME?: string;
+      ALLOWED_IFRAME_SOURCES?: string;
+      CUSTOM_VISUALIZERS?: string;
+      CUSTOM_VISUALIZERS_TYPES?: string;
```
Comment (IlyaBondar) [inline apps/chat/environment.d.ts]: These new variables should be mentioned in chat README with links to `libs/visualizer-connector/README.md`
Also chat README should have section/link for custom renderers as we hae for overlay

File: libs/visualizer-connector/README.md [hunk 22]
```diff
@@ -0,0 +1,193 @@
...
+ALLOWED_IFRAME_SOURCES=http://localhost:8000
+```
+
+Moreover, it needs to be configured some **Visualizer** properties:
+
+- `CUSTOM_VISUALIZERS_TYPES` - list of the attachment content types for which you want to use custom visualizers.
+
+```
+CUSTOM_VISUALIZERS_TYPES='<custom_type_1>, <custom_type_2>'
+```
```
Comment (IlyaBondar) [inline libs/visualizer-connector/README.md]: I didn't understand for what do we need this list.
`CUSTOM_VISUALIZERS` already have `ContentType` property.

File: apps/chat/src/components/Chat/MessageAttachment.tsx [hunk 23]
```diff
@@ -326,9 +385,9 @@ export const MessageAttachment = ({ attachment, isInner }: Props) => {
             (attachment.type === chartType ? (
               <ChartAttachmentUrlRenderer attachmentUrl={mappedAttachmentUrl} />
             ) : (
-              <AttachmentUrlRenderer
-                attachmentUrl={mappedAttachmentUrl}
+              <AttachmentUrlRendererComponent
                 attachmentType={attachment.type}
+                mappedAttachmentUrl={mappedAttachmentUrl}
```
Comment (IlyaBondar) [inline apps/chat/src/components/Chat/MessageAttachment.tsx]: why was it renamed from `attachmentUrl` to `mappedAttachmentUrl`?

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1297.nightly-test.deltixhub.io
>E2E tests status: failed

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1297.nightly-test.deltixhub.io
>E2E tests status: success

File: libs/visualizer-connector/src/lib/VisualizerConnector.ts [hunk 24]
```diff
@@ -0,0 +1,318 @@
...
+  }
+
+  /**
+   * Destroys Visualizer
+   */
+  destroy() {
+    window.removeEventListener('message', this.process);
+    this.iframeInteraction.fail('Chat Visualizer destroyed');
+    this.root.removeChild(this.iframe);
+    //TODO check do we need same changes in Overlay lib
```
Comment (IlyaBondar) [inline libs/visualizer-connector/src/lib/VisualizerConnector.ts]: Did you check?

File: libs/visualizer-connector/package.json [hunk 25]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "@epam/ai-dial-visualizer-connector",
+  "description": "Package for connecting custom visualizer into Dial Chat",
+  "homepage": "https://epam-rail.com",
+  "version": "0.4.3",
```
Comment (IlyaBondar) [inline libs/visualizer-connector/package.json]: Why do we already have version `0.4.3`?

File: libs/chat-visualizer-connector/package.json [hunk 26]
```diff
@@ -0,0 +1,19 @@
+{
+  "name": "@epam/ai-dial-chat-visualizer-connector",
+  "description": "Package for custom visualizer to connect with Dial Chat",
+  "homepage": "https://epam-rail.com",
+  "version": "0.4.3",
```
Comment (IlyaBondar) [inline libs/chat-visualizer-connector/package.json]: Why do we already have version 0.4.3?

File: apps/chat/src/components/Chat/MessageAttachment.tsx [hunk 27]
```diff
@@ -326,9 +364,9 @@ export const MessageAttachment = ({ attachment, isInner }: Props) => {
             (attachment.type === chartType ? (
```
Comment (IlyaBondar) [inline apps/chat/src/components/Chat/MessageAttachment.tsx]: @denys-kolomiitsev , custom renderer should has higher priority than even Plotly

**Comment by denys-kolomiitsev** [comment]
/deploy-review
>Environment URL: https://chat-ai-dial-chat-pr-1297.nightly-test.deltixhub.io
>E2E tests status: success

**Review by IlyaBondar — DISMISSED** [review]

File: README.md [hunk 28]
```diff
@@ -18,6 +18,8 @@ This repository is managed as monorepo by [NX](https://nx.dev/) tools.
 
 `DIAL Overlay` documentation is placed [here](./libs/overlay/README.md).
 
+`DIAL Chat Visualizer Connector` documentation is placed [here](./libs/chat-visualizer-connector/README.md).
+
```
Comment (mikitabut) [inline README.md:22]: Add link to visualizer connector also here

File: apps/chat/src/types/custom-visualizers.ts [hunk 29]
```diff
@@ -0,0 +1,9 @@
+export interface CustomVisualizer {
+  Title: string;
```
Comment (mikitabut) [inline apps/chat/src/types/custom-visualizers.ts]: why is pascal case notation here? We are using camelCase in entire code base

**Review by IlyaBondar — APPROVED** [review]

**Review by mikitabut — APPROVED** [review]

**Comment by sr-remsha** [comment]
Documented in [user guide](https://docs.epam-rail.com/user-guide#data-visualization)

**Pull request merged**